File: 백준/Gold/2096. 내려가기/내려가기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        
        int[][] score = new int[n][3];
        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < 3; j++) {
                score[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int[][] dpMax = new int[n + 1][3];
        for (int i = 1; i <= n; i++) {
            dpMax[i][0] = Math.max(dpMax[i-1][0], dpMax[i-1][1]) + score[i-1][0];
            dpMax[i][1] = Math.max(Math.max(dpMax[i-1][0], dpMax[i-1][1]), dpMax[i-1][2]) + score[i-1][1];
            dpMax[i][2] = Math.max(dpMax[i-1][1], dpMax[i-1][2]) + score[i-1][2];
        }

        int[][] dpMin = new int[n + 1][3];
        for (int i = 1; i <= n; i++) {
            dpMin[i][0] = Math.min(dpMin[i-1][0], dpMin[i-1][1]) + score[i-1][0];
            dpMin[i][1] = Math.min(Math.min(dpMin[i-1][0], dpMin[i-1][1]), dpMin[i-1][2]) + score[i-1][1];
            dpMin[i][2] = Math.min(dpMin[i-1][1], dpMin[i-1][2]) + score[i-1][2];
        }

        System.out.print(Math.max(Math.max(dpMax[n][0], dpMax[n][1]), dpMax[n][2]));
        System.out.print(" ");
        System.out.print(Math.min(Math.min(dpMin[n][0], dpMin[n][1]), dpMin[n][2]));
    }
}

File: 백준/Gold/2096. 내려가기/내려가기.cc
#include <iostream>
#include <algorithm>

using namespace std;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	int n;
	cin >> n;

	int score[3];
	int dp_max[2][3];
	int dp_min[2][3];

	fill_n(dp_max[0], 3, 0);
	fill_n(dp_min[0], 3, 0);

	for (int i = 0; i < n; i++) {
		cin >> score[0] >> score[1] >> score[2];

		dp_max[1][0] = max(dp_max[0][0], dp_max[0][1]) + score[0];
		dp_max[1][1] = max(max(dp_max[0][0], dp_max[0][1]), dp_max[0][2]) + score[1];
		dp_max[1][2] = max(dp_max[0][1], dp_max[0][2]) + score[2];

		dp_min[1][0] = min(dp_min[0][0], dp_min[0][1]) + score[0];
		dp_min[1][1] = min(min(dp_min[0][0], dp_min[0][1]), dp_min[0][2]) + score[1];
		dp_min[1][2] = min(dp_min[0][1], dp_min[0][2]) + score[2];

		for (int j = 0; j < 3; j++) {
			dp_max[0][j] = dp_max[1][j];
			dp_min[0][j] = dp_min[1][j];
		}
	}

	cout << max(max(dp_max[1][0], dp_max[1][1]), dp_max[1][2]) << " ";
	cout << min(min(dp_min[1][0], dp_min[1][1]), dp_min[1][2]);

	return 0;
}